#### **Valuation of Fixed Income Securities**

**Present Value**: The value of future cash flows discounted at the required rate of return.

**Yield to Maturity (YTM)**: The internal rate of return earned by an investor who buys the bond today and holds it until maturity.

**Bond Pricing Formula**:

$$
P = \sum_{t=1}^{n} \frac{C}{(1 + r)^t} + \frac{F}{(1 + r)^n}
$$

- $P$: Price of the bond
- $C$: Periodic coupon payment
- $r$: Discount rate (YTM)
- $F$: Face value of the bond
- $n$: Number of periods


In [1]:
# Example data
coupon_payment = 50  # Periodic coupon payment
face_value = 1000    # Face value of the bond
ytm = 0.05           # Yield to maturity (5%)
periods = 10         # Number of periods

# Calculate bond price
bond_price = sum([coupon_payment / (1 + ytm)**t for t in range(1, periods + 1)]) + face_value / (1 + ytm)**periods
print(f"Bond Price: ${bond_price:.2f}")


Bond Price: $1000.00


#### **Duration and Convexity**

**Duration**: Measures the sensitivity of a bond's price to changes in interest rates.

- **Modified Duration**:

$$
D_{\text{mod}} = \frac{D}{1 + \frac{r}{m}}
$$

  - $D$: Macaulay duration
  - $r$: Yield to maturity
  - $m$: Number of compounding periods per year

**Convexity**: Measures the curvature of the price-yield relationship of a bond.

- **Convexity Formula**:

$$
\text{Convexity} = \frac{\sum_{t=1}^{n} \frac{t(t+1)C}{(1+r)^{t+2}} + \frac{n(n+1)F}{(1+r)^{n+2}}}{P}
$$


In [2]:
# Example data
coupon_payment = 50
face_value = 1000
ytm = 0.05
periods = 10
m = 1  # Compounding periods per year

# Calculate Macaulay duration
cash_flows = [coupon_payment] * periods + [face_value]
discount_factors = [(1 + ytm)**t for t in range(1, periods + 1)]
present_values = [cf / df for cf, df in zip(cash_flows, discount_factors)]
weighted_average = sum([(t + 1) * pv for t, pv in enumerate(present_values[:-1])] + [(periods + 1) * present_values[-1]])
bond_price = sum(present_values)
macaulay_duration = weighted_average / bond_price

# Calculate modified duration
modified_duration = macaulay_duration / (1 + ytm / m)
print(f"Modified Duration: {modified_duration:.4f}")

# Calculate convexity
convexity = sum([(t + 1) * (t + 2) * pv / ((1 + ytm)**2) for t, pv in enumerate(present_values[:-1])] + [(periods * (periods + 1) * present_values[-1] / ((1 + ytm)**2))]) / bond_price
print(f"Convexity: {convexity:.4f}")


Modified Duration: 4.9320
Convexity: 35.6023


### Interpretation of Results

The calculated values for **Modified Duration** and **Convexity** provide insights into the sensitivity of the bond's price to changes in interest rates:

1. **Modified Duration**: 
   - The modified duration value of **4.9320** indicates that for every 1% change in interest rates, the price of the bond is expected to change by approximately 4.9320% in the opposite direction. 
   - A higher duration means the bond is more sensitive to interest rate changes, and thus, it carries more interest rate risk.

2. **Convexity**:
   - The convexity value of **35.6023** shows the degree to which the duration of the bond changes as interest rates change.
   - Positive convexity implies that as interest rates decrease, the bond price increases at an increasing rate, and as interest rates increase, the bond price decreases at a decreasing rate.
   - Convexity is an important measure to assess the bond’s price volatility beyond what is captured by duration alone.

These metrics are crucial for fixed-income investors to evaluate the risk and potential price volatility of bonds in response to interest rate fluctuations.


#### **Analysis of Fixed Income Securities**

**Credit Risk**: The risk that the bond issuer will default on interest or principal payments.

**Interest Rate Risk**: The risk that changes in interest rates will affect bond prices.

**Inflation Risk**: The risk that inflation will erode the purchasing power of bond payments.

**Reinvestment Risk**: The risk that cash flows from a bond will be reinvested at lower rates.


In [3]:
%pip install -q yfinance

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [7]:
import yfinance as yf

# Example data: Bond issuer's stock ticker
issuer_ticker = 'AAPL'

# Fetch financial data
issuer_data = yf.Ticker(issuer_ticker)
balance_sheet = issuer_data.balance_sheet.T  # Transpose the DataFrame
income_statement = issuer_data.financials.T  # Transpose the DataFrame
cash_flow_statement = issuer_data.cashflow.T  # Transpose the DataFrame

# Ensure that the necessary fields exist, excluding the 2019-09-30 period
total_debt = balance_sheet.get('Total Debt').dropna()
stockholders_equity = balance_sheet.get('Stockholders Equity').dropna()
current_assets = balance_sheet.get('Current Assets').dropna()
current_liabilities = balance_sheet.get('Current Liabilities').dropna()
ebit = income_statement.get('EBIT').dropna()
interest_expense = income_statement.get('Interest Expense').dropna()

# Key financial ratios with NaN handling, excluding periods with missing data
debt_to_equity = total_debt / stockholders_equity if not total_debt.empty and not stockholders_equity.empty else None
interest_coverage = ebit / interest_expense if not ebit.empty and not interest_expense.empty else None
current_ratio = current_assets / current_liabilities if not current_assets.empty and not current_liabilities.empty else None

# Print financial ratios (using the latest available data)
print(f"Debt-to-Equity Ratio: {debt_to_equity[-1]:.2f}" if debt_to_equity is not None else "Debt-to-Equity Ratio: Data Unavailable")
print(f"Interest Coverage Ratio: {interest_coverage[-1]:.2f}" if interest_coverage is not None else "Interest Coverage Ratio: Data Unavailable")
print(f"Current Ratio: {current_ratio[-1]:.2f}" if current_ratio is not None else "Current Ratio: Data Unavailable")


Debt-to-Equity Ratio: 1.87
Interest Coverage Ratio: 24.35
Current Ratio: 1.36


### Interpretation of Financial Ratios

The financial ratios calculated provide key insights into the financial health and risk profile of the bond issuer (in this case, Apple Inc.):

1. **Debt-to-Equity Ratio**: 
   - The calculated Debt-to-Equity Ratio is **1.87**, indicating that the company has $1.87 in debt for every $1 of equity. 
   - This ratio suggests a relatively high level of leverage, meaning the company relies significantly on debt financing. Investors should consider this ratio in the context of the industry average and the company's ability to service its debt.

2. **Interest Coverage Ratio**:
   - The Interest Coverage Ratio is **24.35**, which means that the company earns 24.35 times its interest obligations from its operating earnings (EBIT).
   - A high interest coverage ratio indicates that the company is easily able to meet its interest payments, which is a positive sign for bondholders concerned with the issuer's ability to meet its debt obligations.

3. **Current Ratio**:
   - The Current Ratio is **1.36**, indicating that the company has $1.36 in current assets for every $1 of current liabilities.
   - A Current Ratio above 1 suggests that the company has a good level of liquidity, meaning it can cover its short-term obligations with its short-term assets. This is a healthy liquidity position, reducing the risk for bondholders.

These ratios provide a snapshot of the company’s financial stability and ability to meet its debt obligations, which are crucial factors for assessing the risk associated with the bond.
